URL Encoding Updates

ben-jones opened the pull request:
Two main pieces of functionality here
1. Can now encode urls so they look they Baidu search queries
2. Bug fixes for market url encoding and changed way that padding is handled

ben-jones committed:
Bug fixes

Fixed up the encode and decode as market functions to properly do
padding. Fixed all other issues that the unit tests caught while
testing

ben-jones committed:
Added unit test for isBaidu()

Added unit test for isBaidu function and and fixed relevant bugs

ben-jones committed:
Bug fixes for Baidu encoding

Fixed a few bugs related to the Baidu encoding scheme. I wrote
the unit tests for the functionality and made adjustments so that
everything passed correctly

ben-jones committed:
More bug fixes for Baidu encoding

Added a few more bug fixes for the Baidu encoding scheme and
cleaned up some whitespace.

ben-jones committed:
Added unit tests for encoding as English

Added unit tests for the encodeAsEnglish function and its
correlary for decoding, decodeAsEnglish. Also fixed some whitespace
issues

gsathya commented on htpt/urlEncode.py:
```diff
@@ -151,18 +155,28 @@ def encodeAsMarket(data):
+    padding = ''
+  dataLen = hex(len(hexData))[2:]
+  if len(dataLen) == 1:
+    dataLen = '0' + dataLen
```
What does this do?

ben-jones replied:
it pads the data length field. If the data is shorter than 16 characters, then the data field only has 1 nibble, and we need the length to always be 2 hex character, aka 1 byte.

gsathya replied:
Please add a comment explaining this. It's not terribly obvious

gsathya commented on htpt/urlEncode.py:
```diff
@@ -179,6 +193,107 @@ def isMarket(url):
+  #values and we are using uppercase to distinguish padding and
+  #actual text
+  url = 'http://www.baidu.com/s?wd=' + urlData
+  retVal = {'url':url, 'cookie':cookies}
```
Please use something like `encodedData` instead of `retVal`

gsathya commented on htpt/urlEncode.py:
```diff
@@ -188,12 +303,12 @@ def decodeAsMarket(url):
   data = matches.group('hash')
   # strip any padding
-  data = data.strip('ABCDEF')
+  dataLen = int(data[:2], 16)
```
why 16? why 2?

ben-jones replied:
This is a one liner to convert the included length info into an integer value that I can use. The 16 tells python what numbering system this is coming from, in this case, it is telling python to interpret the characters as hex.

gsathya replied:
Please add a comment explaining this?

ben-jones committed:
Code review changes

Renamed retVal to encodedData to be more informative

gsathya commented on tests/verifyUrlEncode.py:
```diff
@@ -129,8 +130,9 @@ def test_encodeAsMarket(self):
         self.assertEqual(len(hashPart), 80)
-        hashPart = hashPart.rstrip('ABCDEF')
-        testDatum = binascii.unhexlify(hashPart)
+        dataLen = int(hashPart[:2], 16)
```
Please add a comment here as well :)

gsathya replied:
Or not .. since this is just a test. Clearly I need to sleep soon

gsathya commented:
Hey @ben-jones, can you add the comments so that I can merge this? Thanks

ben-jones committed:
Added explanatory comments

Added comments to two sections of the code to explain why and how
I am converting a hex string into an integer for later use in the
program

gsathya commented:
Merged. Thanks! :)
